### PR TITLE
User API's PR

### PR DESCRIPTION
**Pull request by vikas143516**
Implemented below Api's
+ Add user
+ Edit user
+ List users

### CODE REVIEW

**Comment by bharathkandula99** [comment]
image
Try to fix these

File: src/controllers/user/users.middleware.js [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+
+exports.validateGetUsersInput = async (req, res, next) => {
+  let errorsList = [];
+  try {
+    const { search, pageLimit, pageNumber } = req.query;
+    if (search?.trim() !== ' ' && typeof search !== 'string') {
+      errorsList.push(messages.errorMessages.INVAILD_SEARCH_KEY);
+    }
+    if (pageLimit && (pageLimit < 0 || typeof pageLimit !== 'number')) {
+      errorsList.push(messages.errorMessages.PAGE_LIMIT_MESSAGE);
```
Comment (naveen-repos) [inline src/controllers/user/users.middleware.js:185]: Page Limit and Page Number should be mandatory params. Where user should be paginated.

File: src/controllers/user/users.middleware.js [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+      user_type,
+    } = req.body;
+    const errorsList = [];
+    if (!first_name?.trim() || typeof first_name !== 'string') {
+      errorsList.push(
+        messages.errorMessages.INVAILD_STRING_OR_MISSING_ERROR('first_name'),
+      );
+    }
+    if (user_type && (!first_name?.trim() || typeof first_name !== 'string')) {
+      errorsList.push(messages.errorMessages.INVALID_USER_TYPE_MESSAGE);
```
Comment (naveen-repos) [inline src/controllers/user/users.middleware.js:29]: Didn't get this validation. Do you want to check the user_type?

**Review by naveen-repos — CHANGES_REQUESTED** [review]
In the error response function we don't need that many or conditions. It is in edfora multiple people wrote that for years so they that many or conditions

As we are starting we need to maintain single formate.

File: src/controllers/user/users.helper.js [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+      email: user.email,
+      phone_number: user.phone_number,
+      country_code: user.country_code,
+      address: user.address,
+      pin_code: user.pin_code,
+      state: user.state,
+      user_type: user.user_type || 'USER',
+    };
+    const userData = await aergov_users.create(params).catch((err) => {
+      logger.error(err);
```
Comment (naveen-repos) [inline src/controllers/user/users.helper.js:23]: We don't need catch here. As you have catch for whole function it handle it. And your doing only single job that can take care

File: src/controllers/user/users.helper.js [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+        where: { id },
+        returning: true,
+      })
+      .catch((err) => {
+        return {
+          success: false,
+          data: err,
+        };
+      });
+    if (!userData.success) {
```
Comment (naveen-repos) [inline src/controllers/user/users.helper.js:57]: Please check this code

File: src/controllers/user/users.helper.js [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+      })
+      .then((data) => {
+        return data[0];
+      })
+      .catch((err) => {
+        return undefined;
+      });
+    return data;
+  } catch (err) {
+    return false;
```
Comment (naveen-repos) [inline src/controllers/user/users.helper.js:89]: When it is a helper function please maintain the response formate unique

File: src/controllers/user/users.helper.js [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+};
+
+exports.getUsersListHelper = async (search_key, pageLimit, pageNumber) => {
+  try {
+    const query = getListUsersQuery(search_key, pageLimit, pageNumber);
+    let limit = pageLimit || 10;
+    const data = await sequelize
+      .query(query)
+      .then((data) => {
+        return {
```
Comment (naveen-repos) [inline src/controllers/user/users.helper.js:100]: Can we please avoid these then and catch. Please stick with async and await only through out the project

File: src/controllers/user/users.helper.js [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+    if (data.error) {
+      throw data.message;
+    }
+    return {
+      success: true,
+      data: data,
+    };
+  } catch (err) {
+    return {
+      success: false,
```
Comment (naveen-repos) [inline src/controllers/user/users.helper.js:126]: Log the error occurred

Following helper function also don't have logging of exception occurred in catch block. Please for those also

File: src/controllers/user/user.helper.js [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+const {
+  getDataById,
+  getListUsersQuery,
+  getUserRoleId,
+} = require('./user.query');
+
+exports.addUserHelper = async (user) => {
+  const transaction = await sequelize.transaction();
+  try {
+    if (!user.user_type) user.user_type = 'USER' + user.email;
```
Comment (naveen-repos) [inline src/controllers/user/user.helper.js:17]: No this wrong.

File: src/controllers/user/user.helper.js [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+    let totalPages = Math.round(
+      parseInt(data[0][0]?.data_count || 0) / page_limit,
+    );
+    return {
+      success: true,
+      data: {
+        users: data[0],
+        page_limit: parseInt(page_limit) || 10,
+        page_number: parseInt(page_number) || 1,
+        totalPages: totalPages !== 0 ? totalPages : 1,
```
Comment (naveen-repos) [inline src/controllers/user/user.helper.js:141]: The variable in the response should be in the snake case

File: src/controllers/user/user.middleware.js [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+    const errorsList = [];
+    if (!id?.trim() || typeof id !== 'string' || !id.startsWith('u')) {
+      errorsList.push(messages.errorMessages.INVAILD_USER_ID_MESSAGE);
+    }
+    if (id.startsWith('u')) {
+      const user = await validateDataInDBById(id, dbTables.USERS_TABLE);
+      if (!user.data || !user.success) {
+        errorsList.push(messages.errorMessages.INVAILD_USER_ID_MESSAGE);
+      }
+    }
```
Comment (naveen-repos) [inline src/controllers/user/user.middleware.js:117]: `!id.startsWith('u')` you're doing check shre and again checking at line `112`

File: src/controllers/user/user.middleware.js [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+    if (
+      role_id &&
+      (!role_id?.trim() ||
+        typeof role_id !== 'string' ||
+        !role_id?.startsWith('r'))
+    ) {
+      errorsList.push(messages.errorMessages.INVALID_ROLE_FOUND);
+    }
+    if (role_id?.startsWith('r')) {
```
Comment (naveen-repos) [inline src/controllers/user/user.middleware.js:157]: duplication check

File: src/controllers/user/user.middleware.js [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+      throw errorsList.join(' ,');
+    }
+    return next();
+  } catch (error) {
+    logger.error(error);
+    return errorResponse({
+      req,
+      res,
+      error,
+      message: error,
```
Comment (naveen-repos) [inline src/controllers/user/user.middleware.js:183]: For an instance of string you're sending instance of the exception.

File: src/controllers/user/user.helper.js (hunk 8 above)
Comment (naveen-repos) [reply src/controllers/user/user.helper.js:17]: The user_type = "USER" only.

**Review by naveen-repos — APPROVED** [review]

**Pull request merged**